lib/rspec-puppet/support.rb

```diff
@@ -217,7 +217,23 @@ def with_vardir
       end
     end
 
+    def reset_providers
+      Puppet::Type.eachtype do |type|
+        # Unset the default provider for each type as this is cached in a class
+        # instance method and so doesn't get re-evaluated when we change facts
+        # and compile a new catalogue.
+        type.defaultprovider = nil
+      end
+    end
+
     def build_catalog_without_cache(nodename, facts_val, trusted_facts_val, hiera_config_val, code, exported)
+      # Blindly stub out any feature and command Puppet::Confine evaluations
+      # that providers might do. This is necessary because we convert the
+      # resource catalogue to a RAL catalogue in order to evaluate all the
+      # automatic graph edges, which requires evaluating the providers.
+      allow_any_instance_of(Puppet::Confine::Feature).to receive(:pass?).with(anything).and_return(true)
```

Comment:
I'm hesitant to throw away compatibility if all we buy is one less `if` statement. We could alternatively do a 2.99 release similar to how rspec did it, which contains most of the new API and deprecation warnings, and then drop all old shim code when switching over to 3.0 (or 5.0 to match up with puppet's new platform versioning).

A simple workaround would be to guard the whole block here on `Puppet.version > '4'`, and fix it in documentation. It's not like we need to contort ourselves *too* much to keep compatibility.